packages/tool-cache/src/manifest.ts

```diff
@@ -0,0 +1,96 @@
+import * as semver from 'semver'
+import {debug} from '@actions/core'
+
+// needs to be require for core node modules to be mocked
+// eslint-disable-next-line @typescript-eslint/no-require-imports
+import os = require('os')
+
+/*
+NOTE: versions must be sorted descending by version in the manifest
+      this library short circuits on first semver spec match
+[
+  {
+    "version": "1.2.3",
+    "stable": true,
+    "release_url": "https://github.com/actions/sometool/releases/tag/1.2.3-20200402.6",
+    "files": [
+      {
+        "filename": "sometool-1.2.3-linux-x64.zip",
+        "arch": "x64",
+        "platform": "linux",
+        "download_url": "https://github.com/actions/sometool/releases/tag/1.2.3-20200402.6/sometool-1.2.3-linux-x64.zip"
+      },
+    ...
+    ]
+  },
+  ...
+]
+*/
+
+export interface IToolReleaseFile {
+  filename: string
+  // 'aix', 'darwin', 'freebsd', 'linux', 'openbsd',
+  // 'sunos', and 'win32'
+  // in some images built for hosted it can also be suffixed by version
+  // e.g. linux-1804
+  platform: string
+
+  // 'arm', 'arm64', 'ia32', 'mips', 'mipsel',
+  // 'ppc', 'ppc64', 's390', 's390x',
+  // 'x32', and 'x64'.
+  arch: string
+
+  download_url: string
+}
+
+export interface IToolRelease {
+  version: string
+  stable: boolean
+  release_url: string
+  files: IToolReleaseFile[]
+}
+
+export async function _findMatch(
+  versionSpec: string,
+  stable: boolean,
+  candidates: IToolRelease[]
+): Promise<IToolRelease | undefined> {
+  const archFilter = os.arch()
+  const platFilter = os.platform()
```

Comment:
I think we should pass these values from action itself instead of getting from `os` so different actions will be able to specify these values differently. For example, `platFilter` will be different:
- Python (`macos-1014`, `ubuntu-1604`, `ubuntu-1804`, `windows-2016`)
- Node (`linux`, `darwin`, `windows`) 

For example, currently, we install package `python-3.0.1-macos-1014-x64.tar.gz` to both `macOS 10.14` and `macOS 10.15` images. But if we find any issue with 10.15 tomorrow, we will push additional asset to releases `python-3.0.1-macos-1015-x64.tar.gz` and will modify task to consume it for macOS 10.15.

`platform` value in `versions-manifest.json` doesn't point to target platform where this package will be used. It points to source platform where this package was built. (it was done because the single package can be used on a few platforms).
For example, Node.JS doesn't depend on version of linux, we have the single package for all linux systems. But Python depends. So this mapping `os.platform` -> `versions-manifest.platform` can be different for different tools

If we pass these values from action to toolkit, our code will be more flexible from this side.

Reply:
Yes, I had a TODO to handle this next since I saw that in the python json.

We haven't and shouldn't push the problem to the action author.  All of that information can be read from the environment (and mocked in tests).  Remember the toolkit already reads this from the environment in other places.  For example, when it writes what's downloaded to the cache.

However supporting platforms like `ubuntu-1804` should be straight forward.  We can match on os.platform(), then ImageOS variable (set by hosted env or provisioner) and optionally os.platform() + versionMMmm 

I'll make that change next so it's clearer.

Reply:
Discussed with Maxim.  We are going to split platform and platform_version in the json.  It will match and platform and if optional version filter is specified it will match on that version spec.